Source/WebCore/platform/audio/gstreamer/AudioEncoderGStreamer.cpp

```diff
@@ -0,0 +1,375 @@
+        GST_DEBUG_CATEGORY_INIT(webkit_audio_encoder_debug, "webkitaudioencoder", 0, "WebKit WebCodecs Audio Encoder");
+    });
+
+    GRefPtr<GstElement> element;
+    if (codecName.startsWith("pcm-"_s)) {
+        auto components = codecName.split('-');
+        if (components.size() != 2) {
+            gstEncoderWorkQueue().dispatch([callback = WTFMove(callback), codecName]() mutable {
+                callback(makeUnexpected(makeString("Invalid LPCM codec string: "_s, codecName)));
+            });
+            return;
+        }
+        element = gst_element_factory_make("identity", nullptr);
+    } else {
+        auto& scanner = GStreamerRegistryScanner::singleton();
+        auto lookupResult = scanner.isCodecSupported(GStreamerRegistryScanner::Configuration::Encoding, codecName);
+        if (!lookupResult) {
+            gstEncoderWorkQueue().dispatch([callback = WTFMove(callback), codecName]() mutable {
+                callback(makeUnexpected(makeString("No GStreamer encoder found for codec ", codecName)));
+            });
+            return;
+        }
+        element = gst_element_factory_create(lookupResult.factory.get(), nullptr);
+    }
+    auto encoder = makeUniqueRef<GStreamerAudioEncoder>(codecName, WTFMove(descriptionCallback), WTFMove(outputCallback), WTFMove(postTaskCallback), WTFMove(element));
+    auto error = encoder->initialize(config);
+    gstEncoderWorkQueue().dispatch([callback = WTFMove(callback), encoder = WTFMove(encoder), error]() mutable {
+        auto internalEncoder = encoder->m_internalEncoder;
+        internalEncoder->postTask([callback = WTFMove(callback), encoder = WTFMove(encoder), error]() mutable {
+            if (!error.isEmpty()) {
+                GST_WARNING("Error creating encoder: %s", error.ascii().data());
+                callback(makeUnexpected(makeString("GStreamer encoding initialization failed with error: ", error)));
+                return;
+            }
+
+            GST_DEBUG("Encoder created");
+            callback(UniqueRef<AudioEncoder> { WTFMove(encoder) });
+        });
+    });
+}
+
+GStreamerAudioEncoder::GStreamerAudioEncoder(const String& codecName, DescriptionCallback&& descriptionCallback, OutputCallback&& outputCallback, PostTaskCallback&& postTaskCallback, GRefPtr<GstElement>&& element)
+    : m_internalEncoder(GStreamerInternalAudioEncoder::create(codecName, WTFMove(descriptionCallback), WTFMove(outputCallback), WTFMove(postTaskCallback), WTFMove(element)))
+{
+}
+
+GStreamerAudioEncoder::~GStreamerAudioEncoder()
+{
+    GST_DEBUG_OBJECT(m_internalEncoder->harness()->element(), "Destroying");
+    close();
+}
+
+String GStreamerAudioEncoder::initialize(const AudioEncoder::Config& config)
+{
+    return m_internalEncoder->initialize(config);
+}
+
+void GStreamerAudioEncoder::encode(RawFrame&& frame, EncodeCallback&& callback)
+{
+    gstEncoderWorkQueue().dispatch([frame = WTFMove(frame), encoder = m_internalEncoder, callback = WTFMove(callback)]() mutable {
```

Comment:
I would think callback should be called on the right thread, executing it in the encoder worker queue might not be good.

Reply:
Hmm, I'm not sure which thread you're talking about :)

Reply:
I think we usually call postTask callback to hop to the right thread and inside call the callback.

Reply:
I'll give it a try... If that works the same change will be needed in the gst video encoder I think. Thanks for the suggestion!